serialise/src/Codec/Serialise/Internal/GeneralisedUTF8.hs

```diff
@@ -0,0 +1,158 @@
+{-# LANGUAGE BangPatterns #-}
+
+module Codec.Serialise.Internal.GeneralisedUTF8
+    ( encodeGenUTF8
+    , UTF8Encoding(..)
+    , decodeGenUTF8
+      -- * Utilities
+    , isSurrogate
+    , isValid
+    ) where
+
+import Control.Monad.ST
+import Data.Bits
+import Data.Char
+import Data.Word
+import qualified Codec.CBOR.ByteArray.Sliced as BAS
+import Data.Primitive.ByteArray
+
+data UTF8Encoding = ConformantUTF8 | GeneralisedUTF8
+                  deriving (Show, Eq)
+
+-- | Is a 'Char' a UTF-16 surrogate?
+isSurrogate :: Char -> Bool
+isSurrogate c = (ord c .&. 0xd800) == 0xd800
+
+-- | Encode a string as (generalized) UTF-8. In addition to the encoding, we
+-- return a flag indicating whether the encoded string contained any surrogate
+-- characters, in which case the output is generalized UTF-8.
+encodeGenUTF8 :: String -> (BAS.SlicedByteArray, UTF8Encoding)
+encodeGenUTF8 st = runST $ do
+    ba <- newByteArray (length st)
+    go ba ConformantUTF8 0 st
+  where
+    go :: MutableByteArray s -> UTF8Encoding
+       -> Int -> [Char]
+       -> ST s (BAS.SlicedByteArray, UTF8Encoding)
+    go ba !enc !off  [] = do
+        ba' <- unsafeFreezeByteArray ba
+        return (BAS.SBA ba' 0 off, enc)
+    go ba enc off  (c:cs)
+      | off + 4 >= cap = do
```

Comment:
If we're going to do this, then we ought to start with `length st + 4` (or similar, like +4 and round up to next word size) as otherwise we'll always end up copying & doubling at the end for the ASCII-only case.

Reply:
Oh dear, good catch.